clifford/test/test_dpga.py

```diff
@@ -0,0 +1,198 @@
+
+import numpy as np
+import pytest
+from clifford.dpga import *
+
+
+class TestBasicDPGA:
+    def test_non_orthogonal_metric(self):
+        for i, a in enumerate(wbasis):
+            for j, b in enumerate(wbasis):
+                res = a|b
+                if i != j and i%4 == j%4:
+                    assert res == (0.5) + 0*w1
+                else:
+                    assert res == 0*w1
+                print(a|b, end=' ')
+            print()
+
+    def test_bivector_identities(self):
+        """
+        These come from section 2 of the paper:
+        R(4, 4) As a Computational Framework for 3-Dimensional Computer Graphics
+        by Ron Goldman and Stephen Mann
+        """
+        for wi in wlist:
+            for wj in wlist:
+                assert wi^wj == -wj*wi
+
+        for wis in wslist:
+            for wjs in wslist:
+                assert wis^wjs == -wjs*wis
+
+        for w in wbasis:
+            assert w**2 == 0*w1
+
+        for wi, wis in zip(wlist, wslist):
+            assert wi*wis == 1 - wis*wi
+
+    def test_up_down(self):
+        rng = np.random.default_rng()  # can pass a seed here later
+        for i in range(1000):
+            p = rng.standard_normal(3)
+            dpga_pnt = up(p)
+            pnt_down = down(np.random.rand()*dpga_pnt)
+            np.testing.assert_allclose(pnt_down, p)
+
+    def test_translate(self):
+        rng = np.random.default_rng()   # can pass a seed here later
+        for i in range(100):
+            tvec = rng.standard_normal(3)
+            wt = tvec[0]*w1 + tvec[1]*w2 + tvec[2]*w3
+            biv = w0s*wt
+            Rt = 1 - biv
+            exp_result = np.e**(-biv)
+            assert Rt == exp_result
+
+            assert Rt * w0 * ~Rt == w0 + wt
+            for wi in [w1, w2, w3]:
+                assert Rt * wi * ~Rt == wi
+
+            assert (Rt*~Rt) == 1 + 0*w1
+
+            pnt_vec = rng.standard_normal(3)
+            pnt = up(pnt_vec)
+            res = Rt*pnt*~Rt
+            desired_result = up(pnt_vec+tvec)
+
+            assert up(pnt_vec) + wt == desired_result
+            assert res == desired_result
+
+            print()
+            print(biv)
+            print(tvec/4)
+            print(biv*biv)
+            print()
```

Comment:
Again, best to remove these - they'll slow the tests down, and are likely not meaningful to anyone but you. If you want to keep them, maybe wrap them in `if hugohadfield_debug:`, and set `hugohadfield_debug = False` at the top of the file.